lib/matplotlib/collections.py

```diff
@@ -2016,6 +2016,65 @@ def set_paths(self):
             self._meshWidth, self._meshHeight, self._coordinates)
         self.stale = True
 
+    def set_array(self, A):
+
+        if isinstance(self, QuadMesh) and A is not None:
+            width, height = self._meshWidth, self._meshHeight
+
+            shape = A.shape
+            faulty_data = False
+
+            A_ = A.ravel()
+
+            if self._shading == 'flat':
+                if len(A.shape) == 1:
+                    if not (len(A.ravel()) == height * width or
+                            len(A.ravel()) == (height + 1 * width + 1)):
+                        faulty_data = True
+                elif not (A.shape == (height, width) or
+                        A.shape == (height + 1, width + 1)):
+                    if (len(A_) == height * width or len(A_)
+                            == (height + 1) * (width + 1)):
+                        _api.warn_deprecated(
+                            "3.3", message=f"Dimensions of A {shape} should match "
+                            f"that of X ({width}) and Y ({height}) and is deprecated "
+                            "since %(since)s. This will become an error %(removal)s.")
+                    else:
+                        faulty_data = True
+
+            elif self._shading == 'nearest':
+                if len(A.shape) == 1:
+                    if (len(A.ravel()) != width * height):
+                        faulty_data = True
+                elif (A.shape != (height, width)):
+                    if len(A_) == height * width:
+                        _api.warn_deprecated(
+                            "3.3", message=f"Dimensions of A {shape} should match "
+                            f"that of X ({width}) and Y ({height}) and is deprecated "
+                            "since %(since)s. This will become an error %(removal)s.")
```

Comment:
This is going to be confusing, because the most likely failure is a transpose, and the message above will confusingly say 

```
Dimensions of A (3, 10) should match that of X (3) and Y (10) and is deprecated...
```
which many folks who accidentally pass in transpose may not be aware of our usual and arbitrary indexing rules.

Maybe:
```suggestion
                            "3.3", message=f"for X ({width}) and Y ({height}) expect A to "
                            f"have shape ({height}, {width}).  Passing A ({shape}) is deprecated"
                            "since %(since)s and will become an error %(removal)s.")
```

Reply:
Ah yes, thanks for this